File: packages/pontil_core/src/pontil_core_ffi.mjs
import {
  action_mode,
  ExitCode$isExit,
  ExitCode$isFailure,
  ExitCode$isSuccess,
} from "./pontil/core.mjs";
import { toList } from "./gleam.mjs";

export function setExitCode(value) {
  if (ExitCode$isFailure(value)) {
    setExitCode_(1);
  } else if (ExitCode$isSuccess(value)) {
    setExitCode_(0);
  } else if (ExitCode$isExit(value)) {
    setExitCode_(value[0]);
  }
}

function setExitCode_(value) {
  if (globalThis?.Deno) {
    globalThis.Deno.exitCode = value;
  } else if (globalThis?.process) {
    globalThis.process.exitCode = value;
  }
}

let outputMode = undefined;

export function getOutputMode() {
  if (outputMode === undefined) {
    outputMode = action_mode();
  }
  return outputMode;
}

export function setOutputMode(mode) {
  outputMode = mode;
}

const secrets = new Set();

export function addSecrets(newSecrets) {
  let list = newSecrets;
  while (list.head !== undefined) {
    secrets.add(list.head);
    list = list.tail;
  }
}

export function getSecrets() {
  return toList(Array.from(secrets));
}

export function clearSecrets() {
  secrets.clear();
}

File: packages/pontil_core/src/pontil_core_ffi.erl
-module(pontil_core_ffi).
-export([
    set_exit_code/1,
    get_output_mode/0,
    set_output_mode/1,
    add_secrets/1,
    get_secrets/0,
    clear_secrets/0
]).

set_exit_code(ExitCode) ->
    Code =
        case ExitCode of
            failure -> 1;
            success -> 0;
            {exit, N} -> N
        end,
    erlang:put(pontil_process_exit_code, Code),
    nil.

get_output_mode() ->
    case persistent_term:get(pontil_output_mode, undefined) of
        undefined ->
            Mode = pontil@core:action_mode(),
            persistent_term:put(pontil_output_mode, Mode),
            Mode;
        Mode ->
            Mode
    end.

set_output_mode(Mode) ->
    persistent_term:put(pontil_output_mode, Mode),
    nil.

add_secrets(NewSecrets) ->
    Existing = persistent_term:get(pontil_secrets, []),
    Updated = lists:usort(NewSecrets ++ Existing),
    persistent_term:put(pontil_secrets, Updated),
    nil.

get_secrets() ->
    persistent_term:get(pontil_secrets, []).

clear_secrets() ->
    persistent_term:put(pontil_secrets, []),
    nil.
